In [1]:
import duckdb
import pandas as pd

In [19]:
duckdb.execute("CREATE OR REPLACE TABLE uszips AS SELECT CAST(zip AS INTEGER) AS zip, * EXCLUDE zip FROM '../data/uszips.csv'")
duckdb.execute("CREATE OR REPLACE TABLE vitals AS SELECT Zipcode as zip, * EXCLUDE Zipcode FROM '../data/Vital_Signs__Home_Prices___by_zip_code_20241206.csv'")

In [20]:
# duckdb.execute("SELECT CAST(zip AS INTEGER), * FROM uszips")
duckdb.sql("SELECT * FROM uszips").df()

zip       lat        lng         city state_id   state_name  zcta  \
0        601  18.18027  -66.75266     Adjuntas       PR  Puerto Rico  True   
1        602  18.36075  -67.17541       Aguada       PR  Puerto Rico  True   
2        603  18.45744  -67.12225    Aguadilla       PR  Puerto Rico  True   
3        606  18.16585  -66.93716      Maricao       PR  Puerto Rico  True   
4        610  18.29110  -67.12243       Anasco       PR  Puerto Rico  True   
...      ...       ...        ...          ...      ...          ...   ...   
33779  99923  55.98043 -130.03803        Hyder       AK       Alaska  True   
33780  99925  55.55398 -132.96276      Klawock       AK       Alaska  True   
33781  99926  55.12617 -131.48928   Metlakatla       AK       Alaska  True   
33782  99927  56.33305 -133.60044  Point Baker       AK       Alaska  True   
33783  99929  56.36089 -132.00635     Wrangell       AK       Alaska  True   

      parent_zcta  population  density county_fips            county_name  \
0            None     16834.0    100.9       72001               Adjuntas   
1            None     37642.0    479.2       72003                 Aguada   
2            None     49075.0    551.7       72005              Aguadilla   
3            None      5590.0     48.7       72093                Maricao   
4            None     25542.0    265.7       72011                 Añasco   
...           ...         ...      ...         ...                    ...   
33779        None        25.0      0.6       02198  Prince of Wales-Hyder   
33780        None       920.0      6.6       02198  Prince of Wales-Hyder   
33781        None      1465.0      4.3       02198  Prince of Wales-Hyder   
33782        None        14.0      1.2       02198  Prince of Wales-Hyder   
33783        None      2133.0      0.3       02275               Wrangell   

                                          county_weights  \
0                        {"72001": 98.74, "72141": 1.26}   
1                                         {"72003": 100}   
2                        {"72005": 99.76, "72099": 0.24}   
3        {"72093": 82.27, "72153": 11.66, "72121": 6.06}   
4      {"72011": 96.71, "72099": 2.82, "72083": 0.37,...   
...                                                  ...   
33779                                     {"02198": 100}   
33780                                     {"02198": 100}   
33781                                     {"02198": 100}   
33782                                     {"02198": 100}   
33783                                     {"02275": 100}   

                    county_names_all          county_fips_all  imprecise  \
0                    Adjuntas|Utuado              72001|72141      False   
1                             Aguada                    72003      False   
2                     Aguadilla|Moca              72005|72099      False   
3        Maricao|Yauco|Sabana Grande        72093|72153|72121      False   
4      Añasco|Moca|Las Marías|Aguada  72011|72099|72083|72003      False   
...                              ...                      ...        ...   
33779          Prince of Wales-Hyder                    02198      False   
33780          Prince of Wales-Hyder                    02198      False   
33781          Prince of Wales-Hyder                    02198      False   
33782          Prince of Wales-Hyder                    02198      False   
33783                       Wrangell                    02275      False   

       military             timezone  
0         False  America/Puerto_Rico  
1         False  America/Puerto_Rico  
2         False  America/Puerto_Rico  
3         False  America/Puerto_Rico  
4         False  America/Puerto_Rico  
...         ...                  ...  
33779     False        America/Sitka  
33780     False        America/Sitka  
33781     False   America/Metlakatla  
33782     False        America/Sitka  
33783     False        America/Sitka  

[33784 rows x 18 columns]

In [21]:
duckdb.sql("SELECT * FROM vitals").df()

zip  Year  MedPrice  MedPrice_IA  PercentChngPrice  \
0     94002  1997    406317       723341              0.00   
1     94002  1998    466617       805091              0.15   
2     94002  1999    524475       868198              0.29   
3     94002  2000    650458      1030736              0.60   
4     94002  2001    686650      1032508              0.69   
...     ...   ...       ...          ...               ...   
5077  95688  2014    385850       437247              1.26   
5078  95688  2015    417625       461199              1.45   
5079  95688  2016    439725       471434              1.58   
5080  95688  2017    464633       482593              1.72   
5081  95688  2018    489967       489967              1.87   

      PercentChngPriceIA  Source  
0                   0.00  Zillow  
1                   0.11  Zillow  
2                   0.20  Zillow  
3                   0.42  Zillow  
4                   0.43  Zillow  
...                  ...     ...  
5077                0.44  Zillow  
5078                0.52  Zillow  
5079                0.55  Zillow  
5080                0.59  Zillow  
5081                0.61  Zillow  

[5082 rows x 7 columns]

In [26]:
duckdb.sql("CREATE TABLE overlap AS SELECT uszips.*, vitals.* FROM uszips JOIN vitals ON uszips.zip = vitals.zip")

In [27]:
duckdb.sql("SELECT * FROM overlap").df()

zip       lat        lng           city state_id  state_name  zcta  \
0     94002  37.51366 -122.29853        Belmont       CA  California  True   
1     94005  37.68820 -122.40873       Brisbane       CA  California  True   
2     94010  37.57306 -122.37828     Burlingame       CA  California  True   
3     94014  37.69070 -122.44769      Daly City       CA  California  True   
4     94015  37.68121 -122.48074      Daly City       CA  California  True   
...     ...       ...        ...            ...      ...         ...   ...   
5077  95492  38.53052 -122.81254        Windsor       CA  California  True   
5078  95497  38.71976 -123.45601  The Sea Ranch       CA  California  True   
5079  95620  38.40599 -121.75689          Dixon       CA  California  True   
5080  95687  38.33440 -121.91765      Vacaville       CA  California  True   
5081  95688  38.41901 -122.02571      Vacaville       CA  California  True   

     parent_zcta  population  density  ... imprecise military  \
0           None       27893   1914.8  ...     False    False   
1           None        4756    400.9  ...     False    False   
2           None       43458   1109.9  ...     False    False   
3           None       46331   2798.9  ...     False    False   
4           None       65125   4446.1  ...     False    False   
...          ...         ...      ...  ...       ...      ...   
5077        None       28096    548.2  ...     False    False   
5078        None        1224     31.0  ...     False    False   
5079        None       22004     35.2  ...     False    False   
5080        None       68996    597.9  ...     False    False   
5081        None       40147    168.5  ...     False    False   

                 timezone  zip_1  Year  MedPrice  MedPrice_IA  \
0     America/Los_Angeles  94002  2018   1773683      1773683   
1     America/Los_Angeles  94005  2018    940950       940950   
2     America/Los_Angeles  94010  2018   2705717      2705717   
3     America/Los_Angeles  94014  2018    930033       930033   
4     America/Los_Angeles  94015  2018   1000300      1000300   
...                   ...    ...   ...       ...          ...   
5077  America/Los_Angeles  95492  1997    194508       346272   
5078  America/Los_Angeles  95497  1997    308458       549130   
5079  America/Los_Angeles  95620  1997    150275       267526   
5080  America/Los_Angeles  95687  1997    142650       253951   
5081  America/Los_Angeles  95688  1997    170608       303724   

     PercentChngPrice  PercentChngPriceIA  Source  
0                3.37                1.45  Zillow  
1                2.98                1.24  Zillow  
2                3.44                1.50  Zillow  
3                3.34                1.44  Zillow  
4                3.03                1.26  Zillow  
...               ...                 ...     ...  
5077             0.00                0.00  Zillow  
5078             0.00                0.00  Zillow  
5079             0.00                0.00  Zillow  
5080             0.00                0.00  Zillow  
5081             0.00                0.00  Zillow  

[5082 rows x 25 columns]

In [28]:
duckdb.sql("SELECT * FROM overlap WHERE state_id = 'FL'").df()

Empty DataFrame
Columns: [zip, lat, lng, city, state_id, state_name, zcta, parent_zcta, population, density, county_fips, county_name, county_weights, county_names_all, county_fips_all, imprecise, military, timezone, zip_1, Year, MedPrice, MedPrice_IA, PercentChngPrice, PercentChngPriceIA, Source]
Index: []

[0 rows x 25 columns]

In [50]:
duckdb.sql("SELECT DISTINCT(city) FROM uszips WHERE state_id = 'FL'").df()

city
0                Day
1         Fort White
2           Hilliard
3    Saint Augustine
4            Raiford
..               ...
462          Oldsmar
463    Crystal Beach
464         Ferndale
465            Gotha
466         Mascotte

[467 rows x 1 columns]

I seem to be finding some discrepancies here in that the cities are numbered to be 467 in number while Wikipedia states that there are only 411 municipalities.
TODO: Figure out what the issue here is?

The case may be that the zipcodes account for areas that are not municipalities?

It seems as though zipcodes are not great mappings considering that zipcodes are allocated by the US Postal Service for delivering mails. There may be more than one zipcode per city/municipality. This is a point that can be added into the slides!

In [33]:
zip_df = duckdb.sql("SELECT * FROM uszips").df()

In [35]:
import math

# Haversine formula to calculate great-circle distance
def haversine(lat1, lon1, lat2, lon2):
    R = 6371  # Radius of Earth in kilometers. Use 3959 for miles.
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)
    
    a = math.sin(delta_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c

# Example: Define center and radius
# TODO: This below portion needs to be retrieved from the model.
center_lat, center_lon = 28.5383, -81.3792  # Orlando, FL
radius_km = 50  # 50 km

# Filter ZIP codes within the radius
zip_df['distance'] = zip_df.apply(
    lambda row: haversine(center_lat, center_lon, row['lat'], row['lng']), axis=1
)
zipcodes_in_radius = zip_df[zip_df['distance'] <= radius_km]

In [39]:
zipcodes_in_radius

zip       lat       lng               city state_id state_name  zcta  \
10775  32701  28.66509 -81.36916  Altamonte Springs       FL    Florida  True   
10777  32703  28.65579 -81.52139             Apopka       FL    Florida  True   
10778  32707  28.66255 -81.31403        Casselberry       FL    Florida  True   
10779  32708  28.68751 -81.27086     Winter Springs       FL    Florida  True   
10780  32709  28.49266 -80.98801          Christmas       FL    Florida  True   
...      ...       ...       ...                ...      ...        ...   ...   
11460  34771  28.26647 -81.11908        Saint Cloud       FL    Florida  True   
11461  34772  28.16534 -81.27273        Saint Cloud       FL    Florida  True   
11464  34786  28.48161 -81.55501         Windermere       FL    Florida  True   
11465  34787  28.48309 -81.62833      Winter Garden       FL    Florida  True   
11467  34797  28.73517 -81.81813             Yalaha       FL    Florida  True   

      parent_zcta  population  density county_fips county_name  \
10775        None     24551.0   1682.7       12117    Seminole   
10777        None     57046.0    685.9       12095      Orange   
10778        None     38868.0   1560.5       12117    Seminole   
10779        None     47437.0    990.0       12117    Seminole   
10780        None      1545.0      4.0       12095      Orange   
...           ...         ...      ...         ...         ...   
11460        None     27537.0     68.8       12097     Osceola   
11461        None     32268.0    143.4       12097     Osceola   
11464        None     45914.0    893.6       12095      Orange   
11465        None     86913.0    499.0       12095      Orange   
11467        None      1304.0    113.0       12069        Lake   

                         county_weights county_names_all county_fips_all  \
10775                    {"12117": 100}         Seminole           12117   
10777     {"12095": 90.8, "12117": 9.2}  Orange|Seminole     12095|12117   
10778                    {"12117": 100}         Seminole           12117   
10779                    {"12117": 100}         Seminole           12117   
10780                    {"12095": 100}           Orange           12095   
...                                 ...              ...             ...   
11460                    {"12097": 100}          Osceola           12097   
11461                    {"12097": 100}          Osceola           12097   
11464                    {"12095": 100}           Orange           12095   
11465  {"12095": 83.13, "12069": 16.87}      Orange|Lake     12095|12069   
11467                    {"12069": 100}             Lake           12069   

       imprecise  military          timezone   distance  
10775      False     False  America/New_York  14.132435  
10777      False     False  America/New_York  19.062680  
10778      False     False  America/New_York  15.210535  
10779      False     False  America/New_York  19.675324  
10780      False     False  America/New_York  38.556906  
...          ...       ...               ...        ...  
11460      False     False  America/New_York  39.508626  
11461      False     False  America/New_York  42.759988  
11464      False     False  America/New_York  18.298573  
11465      False     False  America/New_York  25.104701  
11467      False     False  America/New_York  48.105897  

[93 rows x 19 columns]

In [41]:
duckdb.execute("SELECT DISTINCT(city) FROM zipcodes_in_radius").df()

city
0              Deltona
1               Eustis
2                 Mims
3          Orange City
4              Orlando
5               Yalaha
6             Longwood
7              Sanford
8             Loughman
9             Ferndale
10               Gotha
11         Casselberry
12              Osteen
13           Montverde
14             Oakland
15               Ocoee
16         Saint Cloud
17       Winter Garden
18          Mount Dora
19   Intercession City
20            Clermont
21  Howey In The Hills
22           Kissimmee
23              Geneva
24           Christmas
25           Lake Mary
26              Oviedo
27             Tavares
28         Winter Park
29              Apopka
30              Debary
31            Plymouth
32            Zellwood
33           Davenport
34            Astatula
35   Altamonte Springs
36      Winter Springs
37            Maitland
38            Sorrento
39           Groveland
40          Windermere

Right now, I have a list of the cities that are in the range of effect of the storm.
This is good because given the `R_max`, I will be able to provide the cities that are in immediate effect.
However, I need to find a way to relate these cities to the their residential value datasets.
This could either be via grouping them into countys or finding a dataset that does it by city.

In [60]:
duckdb.execute("CREATE OR REPLACE TABLE metro_zhvi AS SELECT RegionName[:-5] AS region, * FROM '../data/Metro_zhvi_uc_sfrcondo_tier_0.33_0.67_month.csv'")

In [71]:
duckdb.execute("SELECT * FROM metro_zhvi WHERE RegionName='Orlando, FL'").df()

region  RegionID  SizeRank   RegionName RegionType StateName  \
0  Orlando    394943        22  Orlando, FL        msa        FL   

      1996-02-29     1996-03-31     1996-04-30     1996-05-31  ...  \
0  100317.111882  100742.379203  101203.898892  101734.229395  ...   

      2024-01-31     2024-02-29     2024-03-31     2024-04-30    2024-05-31  \
0  388806.888863  390294.466559  393053.179215  395428.929284  396757.65439   

      2024-06-30     2024-07-31     2024-08-31     2024-09-30     2024-10-31  
0  397132.264012  397795.042048  396961.651983  395527.688415  393518.790624  

[1 rows x 351 columns]

In [79]:
duckdb.execute("SELECT DISTINCT(city) FROM zipcodes_in_radius").df()

city
0               Geneva
1              Deltona
2               Eustis
3                 Mims
4          Orange City
5              Orlando
6               Yalaha
7            Christmas
8            Lake Mary
9               Oviedo
10             Tavares
11         Winter Park
12         Casselberry
13              Osteen
14           Montverde
15             Oakland
16               Ocoee
17         Saint Cloud
18       Winter Garden
19   Altamonte Springs
20      Winter Springs
21            Maitland
22            Sorrento
23           Groveland
24          Windermere
25          Mount Dora
26   Intercession City
27            Clermont
28  Howey In The Hills
29           Kissimmee
30              Apopka
31              Debary
32            Plymouth
33            Zellwood
34           Davenport
35            Astatula
36            Longwood
37             Sanford
38            Loughman
39            Ferndale
40               Gotha

TODO: Need to consolidate the different cities that we want to showcase.

We could also showcase the housing price increase and decrease on a yearly basis for cities that fall into the pathway of the storm.

In [70]:
duckdb.execute("SELECT metro_zhvi.* FROM metro_zhvi JOIN zipcodes_in_radius ON metro_zhvi.region = zipcodes_in_radius.city WHERE StateName = 'FL'").df()

region  RegionID  SizeRank   RegionName RegionType StateName  \
0   Orlando    394943        22  Orlando, FL        msa        FL   
1   Deltona    394528        90  Deltona, FL        msa        FL   
2   Orlando    394943        22  Orlando, FL        msa        FL   
3   Deltona    394528        90  Deltona, FL        msa        FL   
4   Orlando    394943        22  Orlando, FL        msa        FL   
5   Orlando    394943        22  Orlando, FL        msa        FL   
6   Orlando    394943        22  Orlando, FL        msa        FL   
7   Orlando    394943        22  Orlando, FL        msa        FL   
8   Orlando    394943        22  Orlando, FL        msa        FL   
9   Orlando    394943        22  Orlando, FL        msa        FL   
10  Orlando    394943        22  Orlando, FL        msa        FL   
11  Orlando    394943        22  Orlando, FL        msa        FL   
12  Orlando    394943        22  Orlando, FL        msa        FL   
13  Orlando    394943        22  Orlando, FL        msa        FL   
14  Orlando    394943        22  Orlando, FL        msa        FL   
15  Orlando    394943        22  Orlando, FL        msa        FL   
16  Orlando    394943        22  Orlando, FL        msa        FL   
17  Orlando    394943        22  Orlando, FL        msa        FL   
18  Orlando    394943        22  Orlando, FL        msa        FL   
19  Orlando    394943        22  Orlando, FL        msa        FL   
20  Orlando    394943        22  Orlando, FL        msa        FL   
21  Orlando    394943        22  Orlando, FL        msa        FL   
22  Orlando    394943        22  Orlando, FL        msa        FL   
23  Orlando    394943        22  Orlando, FL        msa        FL   
24  Orlando    394943        22  Orlando, FL        msa        FL   
25  Orlando    394943        22  Orlando, FL        msa        FL   
26  Orlando    394943        22  Orlando, FL        msa        FL   
27  Orlando    394943        22  Orlando, FL        msa        FL   
28  Orlando    394943        22  Orlando, FL        msa        FL   
29  Orlando    394943        22  Orlando, FL        msa        FL   
30  Orlando    394943        22  Orlando, FL        msa        FL   
31  Orlando    394943        22  Orlando, FL        msa        FL   
32  Orlando    394943        22  Orlando, FL        msa        FL   
33  Orlando    394943        22  Orlando, FL        msa        FL   
34  Orlando    394943        22  Orlando, FL        msa        FL   

       1996-02-29     1996-03-31     1996-04-30     1996-05-31  ...  \
0   100317.111882  100742.379203  101203.898892  101734.229395  ...   
1    88425.915929   88581.588340   88814.294958   89176.673482  ...   
2   100317.111882  100742.379203  101203.898892  101734.229395  ...   
3    88425.915929   88581.588340   88814.294958   89176.673482  ...   
4   100317.111882  100742.379203  101203.898892  101734.229395  ...   
5   100317.111882  100742.379203  101203.898892  101734.229395  ...   
6   100317.111882  100742.379203  101203.898892  101734.229395  ...   
7   100317.111882  100742.379203  101203.898892  101734.229395  ...   
8   100317.111882  100742.379203  101203.898892  101734.229395  ...   
9   100317.111882  100742.379203  101203.898892  101734.229395  ...   
10  100317.111882  100742.379203  101203.898892  101734.229395  ...   
11  100317.111882  100742.379203  101203.898892  101734.229395  ...   
12  100317.111882  100742.379203  101203.898892  101734.229395  ...   
13  100317.111882  100742.379203  101203.898892  101734.229395  ...   
14  100317.111882  100742.379203  101203.898892  101734.229395  ...   
15  100317.111882  100742.379203  101203.898892  101734.229395  ...   
16  100317.111882  100742.379203  101203.898892  101734.229395  ...   
17  100317.111882  100742.379203  101203.898892  101734.229395  ...   
18  100317.111882  100742.379203  101203.898892  101734.229395  ...   
19  100317.111882  100742.379203  101203.898892  101734.229395  ...   
20  100317.111882  100742.379